goldenmaster/apigear/mqtt/private/mqttcwrapper.cpp

```diff
@@ -0,0 +1,593 @@
+    m_onConnectionStatusChangedCallbacksMutex.unlock();
+
+    return subscriptionId;
+}
+
+void CWrapper::unsubscribeToConnectionStatus(int subscriptionID)
+{
+    m_onConnectionStatusChangedCallbacksMutex.lock();
+    if((m_onConnectionStatusChangedCallbacks.find(subscriptionID) != m_onConnectionStatusChangedCallbacks.end()))
+    {
+        m_onConnectionStatusChangedCallbacks.erase(subscriptionID);
+    }
+    m_onConnectionStatusChangedCallbacksMutex.unlock();
+}
+
+void CWrapper::run()
+{
+    do
+    {
+        std::unique_lock<std::mutex> lock{m_waitForSubscriptionChangesMutex};
+        m_synchronizeSubscriptionChanges.wait(lock, [this]() { return !m_waitForSubscriptionChanges; });
+        m_waitForSubscriptionChanges = true;
+        lock.unlock();
+
+        addNewSubscriptions();
+        removeOldSubscriptions();
+
+    }
+    while (m_connected && !m_disconnectRequested);
+
+    unsubscribeAllTopics();
+    waitForPendingMessages();
+}
+
+void CWrapper::addNewSubscriptions()
+{
+    m_toBeSubscribedTopicsMutex.lock();
+    const auto toBeSubscribedTopics(std::move(m_toBeSubscribedTopics));
+    m_toBeSubscribedTopics.clear();
+    m_toBeSubscribedTopicsMutex.unlock();
+    for (const auto& topic : toBeSubscribedTopics) {
+
+        MQTTAsync_responseOptions opts = MQTTAsync_responseOptions_initializer;
+        opts.onSuccess5 = onSubscribeSuccess;
+        opts.onFailure5 = onSubscribeFailure;
+        opts.context = new subscribeTopicContext{topic.first, topic.second, getPtr()};
+        int responseCode = MQTTAsync_subscribe(*m_client.get(), topic.first.c_str(), QOS, &opts);
+        if (responseCode != MQTTASYNC_SUCCESS)
+        {
+            AG_LOG_ERROR("Failed to start subscribe, return code " + std::to_string(responseCode));
+        }
+    }
+}
+
+void CWrapper::removeOldSubscriptions()
+{
+    m_toBeUnsubscribedTopicsMutex.lock();
+    const auto toBeUnsubscribedTopics(std::move(m_toBeUnsubscribedTopics));
+    m_toBeUnsubscribedTopics.clear();
+    m_toBeUnsubscribedTopicsMutex.unlock();
+    for (const auto& topic : toBeUnsubscribedTopics) {
+        AG_LOG_INFO("Unsubscribing from " + topic);
+
+        MQTTAsync_responseOptions opts = MQTTAsync_responseOptions_initializer;
+        opts.onSuccess5 = onUnsubscribeSuccess;
+        opts.onFailure5 = onUnsubscribeFailure;
+        opts.context = new subscribeTopicContext{topic, nullptr, getPtr()};
+        int responseCode = MQTTAsync_unsubscribe(*m_client.get(), topic.c_str(), &opts);
+        if (responseCode != MQTTASYNC_SUCCESS)
+        {
+            AG_LOG_WARNING("Failed to start unsubscribe, return code " + std::to_string(responseCode));
+        }
+    }
+}
+
+void CWrapper::unsubscribeAllTopics()
+{
+    if(!m_connected)
+    {
+        return;
+    }
+
+    // unsubscribe from all topics
+    m_subscribedTopicsMutex.lock();
+    const auto subscribedTopics { m_subscribedTopics };
```

Comment:
std::move?

Reply:
No, that's intentional. It stays in the subscribedTopics map until the unsubscription is confirmed in the `CWrapper::onUnsubscribed` callback